### PR TITLE
Add ranger permission check

### PR DESCRIPTION
**Pull request by lianneli**
add ranger permission check
https://docs.qq.com/doc/DU3dRdHR6UnVnb2Jl?no_promotion=1

### CODE REVIEW

File: sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java [hunk 1]
```diff
@@ -1671,6 +1854,17 @@ public void setPermission(Path p, FsPermission permission) throws IOException {
 
   @Override
   public void setOwner(Path p, String username, String groupname) throws IOException {
```
[the remark this answers is not in the document]
Comment (lianneli) [reply sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java:1847]: > only superuser can do this

Fixed. Please check again.

File: sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java [hunk 2]
```diff
@@ -1412,6 +1546,18 @@ public void concat(final Path dst, final Path[] srcs) throws IOException {
 
   @Override
   public boolean rename(Path src, Path dst) throws IOException {
+    if (needCheckPermission()) {
+      if (!exists(src)) {
+        return false;
+      }
+      checkParentPathAccess(src.getParent(), FsAction.WRITE, "rename");
+      Path dstAncestor = dst.getParent();
+      while (!exists(dstAncestor)) {
```
[the remark this answers is not in the document]
Comment (lianneli) [reply sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java]: > exists may check execute action which is not necessary.

Fixed. Please check again.

File: sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java [hunk 3]
```diff
@@ -478,11 +492,84 @@ public void initialize(URI uri, Configuration conf) throws IOException {
...
     if (!isEmpty(uidFile) || !isEmpty(groupingFile)) {
       updateUidAndGrouping(uidFile, groupingFile);
-      refreshUidAndGrouping(uidFile, groupingFile);
+      if (!isSuperGroupFileSystem) {
+        refreshUidAndGrouping(uidFile, groupingFile);
+      }
+    }
+  }
+
+  private RangerConfig checkAndGetRangerParams(Configuration conf) throws RuntimeException {
```
[the remark this answers is not in the document]
Comment (lianneli) [reply sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java]: > It's better to throw an instance of IOException

Fixed. Please check again.

File: sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java [hunk 4]
```diff
@@ -478,11 +492,84 @@ public void initialize(URI uri, Configuration conf) throws IOException {
...
+        if (!isSuperGroupFileSystem) {
+          RangerConfig rangerConfig = checkAndGetRangerParams(conf);
+          Configuration superConf = new Configuration(conf);
+          superGroupFileSystem = new JuiceFileSystemImpl(true);
+          superGroupFileSystem.initialize(uri, superConf);
+          rangerPermissionChecker = new RangerPermissionChecker(superGroupFileSystem, rangerConfig, getScheme(), name, user, group);
+          permissionCheckerInitSuccess = true;
+        }
+      } catch (Exception e) {
+        LOG.warn("The initialization of the Permission Checker has failed, and permission will be checked using Mask. ", e);
```
[the remark this answers is not in the document]
Comment (lianneli) [reply sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java]: > If ranger init failed, juicefs initialize should fail

Fixed. Please check again.

File: sdk/java/src/main/java/io/juicefs/permission/LockFileChecker.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
+/*
+ * JuiceFS, Copyright 2020 Juicedata, Inc.
```
Comment (tangyoupeng) [inline sdk/java/src/main/java/io/juicefs/permission/LockFileChecker.java:2]: change 2020 to 2024

File: sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java [hunk 6]
```diff
@@ -1550,13 +1549,9 @@ public boolean rename(Path src, Path dst) throws IOException {
       if (!exists(src)) {
         return false;
       }
-      checkParentPathAccess(src.getParent(), FsAction.WRITE, "rename");
-      Path dstAncestor = dst.getParent();
-      while (!exists(dstAncestor)) {
-        dstAncestor = dstAncestor.getParent();
+      if (!checkParentPathAccess(src, FsAction.WRITE, "rename") && !checkAncestorAccess(dst, FsAction.WRITE, "rename")) {
```
[the remark this answers is not in the document]
Comment (lianneli) [reply sdk/java/src/main/java/io/juicefs/JuiceFileSystemImpl.java]: > If checkParentPathAccess(src, FsAction.WRITE, "rename") fallback to defaultfs permission check and passed. It wont check checkAncestorAccess(dst, FsAction.WRITE, "rename").

Fixed. Please check again.

File: sdk/java/src/main/java/io/juicefs/permission/LockFileChecker.java (hunk 5 above)
Comment (lianneli) [reply sdk/java/src/main/java/io/juicefs/permission/LockFileChecker.java]: > change 2020 to 2024

Fixed. Please check again.

**Review by davies — APPROVED** [review]

**Comment by davies** [comment]
@lianneli Thanks for your contribution!

**Pull request merged**